### PR TITLE
collation-generation: Avoid using `para_backing_state` if runtime is ancient

### PR DESCRIPTION
**Pull request by sandreim**
fixes https://github.com/paritytech/polkadot-sdk/issues/4067

Also add an early bail out for look ahead collator such that we don't waste time if a CollatorFn is not set.

TODO:
- [x] add test.
- [x] Polkadot System Parachain burn-in.

### CODE REVIEW

File: polkadot/node/collation-generation/src/lib.rs [hunk 1]
```diff
@@ -301,34 +317,46 @@ async fn handle_new_activations<Context>(
...
-		// For timeout assumption the collator can't really know because it doesn't receive bitfield
-		// gossip.
-		let assumption = if para_backing_state.pending_availability.is_empty() {
-			OccupiedCoreAssumption::Free
-		} else {
-			OccupiedCoreAssumption::Included
+		// If at least one core is assigned to us, `para_assumption` is `Some`.
+		let Some(mut para_assumption) = para_assumption else { continue };
+
+		match async_backing_params {
```
Comment (bkchr) [inline polkadot/node/collation-generation/src/lib.rs:323]: Could be an `if async_backing_params.is_some()`

**Review by bkchr — APPROVED** [review]

**Comment by sandreim** [comment]
Asked for a burn-in to confirm the fix. We'll have to do a hotfix release once this is merged/backported.

**Comment by PierreBesson** [comment]
The burn-in on Polkadot AssetHub collators confirm that this is fixing the issue of collators not producing blocks.

File: polkadot/node/collation-generation/src/lib.rs [hunk 2]
```diff
@@ -209,6 +209,13 @@ async fn handle_new_activations<Context>(
 	if config.collator.is_none() {
 		return Ok(())
 	}
+
+	// If there is no collation function provided, bail out early.
+	// Important: Lookahead collator and slot based collator do not use `CollatorFn`.
+	if config.collator.is_none() {
```
[the remark this answers is not in the document]
Comment (sandreim) [reply polkadot/node/collation-generation/src/lib.rs:215]: I don't think we can do this, the config is wrapped in an Arc.

**Review by bkchr — APPROVED** [review]

**Pull request merged**